MoneyDataType/Amount.cs

```diff
@@ -1,141 +1,149 @@
+[Newtonsoft.Json.JsonConverter(typeof(LegacyAmountConverter))]
+[DebuggerDisplay("{DebuggerDisplay,nq}")]
+public readonly struct Amount : IEquatable<Amount>, IComparable<Amount>
 {
-    /// <summary>
-    /// A money amount, which is represented by a decimal number and a currency
-    /// </summary>
-    [JsonConverter(typeof(AmountConverter))]
-    [Newtonsoft.Json.JsonConverter(typeof(LegacyAmountConverter))]
-    [DebuggerDisplay("{DebuggerDisplay,nq}")]
-    public struct Amount : IEquatable<Amount>, IComparable<Amount>
+    public Amount()
     {
-        public Amount() {
-            Value = 0;
-            Currency = Money.Currency.UnspecifiedCurrency;
-        }
+        Value = 0;
+        Currency = Money.Currency.UnspecifiedCurrency;
+    }
 
-        public Amount(decimal value, RegionInfo region)
-        {
-            if (region == null)
-                throw new ArgumentNullException(nameof(region));
+    public Amount(decimal value, RegionInfo region)
+    {
+        if (region == null) throw new ArgumentNullException(nameof(region));
 
-            Currency = Money.Currency.FromRegion(region);
-            Value = value;
-        }
+        Currency = Money.Currency.FromRegion(region);
+        Value = value;
+    }
+
+    public Amount(decimal value, CultureInfo culture)
+    {
+        if (culture is null) throw new ArgumentNullException(nameof(culture));
 
-        public Amount(decimal value, CultureInfo culture)
-        {
-            if (culture is null)
-                throw new ArgumentNullException(nameof(culture));
+        Currency = Money.Currency.FromCulture(culture);
+        Value = value;
+    }
 
-            Currency = Money.Currency.FromCulture(culture);
-            Value = value;
-        }
+    public Amount(decimal value, ICurrency currency)
+    {
+        Currency = currency ?? throw new ArgumentNullException(nameof(currency));
+        Value = value;
+    }
 
-        /// <summary>
-        /// Constructs a new money amount
-        /// </summary>
-        /// <param name="value">The decimal value</param>
-        /// <param name="currency">The currency</param>
-        public Amount(decimal value, ICurrency currency)
-        {
-            Currency = currency ?? throw new ArgumentNullException(nameof(currency));
-            Value = value;
-        }
+    /// <summary>
+    /// Gets the decimal value.
+    /// </summary>
+    public decimal Value { get; }
 
-        /// <summary>
-        /// The decimal value
-        /// </summary>
-        public decimal Value { get; }
+    /// <summary>
+    /// Gets the currency.
+    /// </summary>
+    public ICurrency Currency { get; }
 
-        /// <summary>
-        /// The currency
-        /// </summary>
-        public ICurrency Currency { get; }
+    public bool Equals(Amount other) =>
+        Value == other.Value &&
+        ((Currency == null && other.Currency == null) || Currency?.Equals(other.Currency) == true);
```

Comment:
`Currency` can't be `null` becuase it's struct yes, but `ICurrency` is interface so it can be `null` even if its only implementation is a value type. `other` can't be null because it's a struct.

Reply:
Very insightful.

Reply:
Because `operator ==` is defined for the type `Currency`, not for `ICurrency`. (I don't think you can define operator overloading for interfaces anyway.) I'm not 100% clear on why we need this `ICurrency` interface and can't just use the `Currency` struct directly, but exactly because I'm not sure I don't dare to touch it.

Reply:
Lord Rider disagrees:
![image](https://user-images.githubusercontent.com/4007293/166078265-4d5a7993-f68e-475d-909e-c008e97ec56f.png)

Reply:
Also if you don't trust our lord and savior (heretic) then here is some empirical proof:
![image](https://user-images.githubusercontent.com/4007293/166079126-5b564d4b-9ab9-42b2-8a76-cf7c35541f48.png)

Reply:
Very enlightening! Thanks for diving into this, my doubts have evaporated.